### ISSUE TITLE
RuntimeWarninRuntimeWarning: invalid value encountered in sqrt   return f_raw(*args, **kwargs)g: invalid value encountered in sqrt   return f_raw(*args, **kwargs)

### ISSUE THREAD

**Issue by sunmouren** [SnapperGPS/snappergps-backend#8]
Aha, it's me again. Last time I used the server you provided to calculate the location and it was displayed normally on the web page, 
![14a931ff-f46e-487a-a806-bf1403a88e3c](https://github.com/user-attachments/assets/7de1ca6e-09dc-4fad-b1bf-af1d0f05d8bf)
but when I built the backend myself, it prompted this when I executed it(I printed some information):

> [('7c02b35559', '24776C025FE526AC', 'waiting', datetime.datetime(2024, 12, 5, 0, 0), datetime.datetime(2024, 12, 13, 11, 10, 10, 833000), '', None, {}, None, None, '')]
Processing record 7c02b35559
1: err -10.000373546116505 Hz, inliers: 1
2: err -20.84699457871088 Hz, inliers: 2
3: err -32.718502326720795 Hz, inliers: 3
4: err -49.070017409930855 Hz, inliers: 4
5: err -49.070017409930855 Hz, inliers: 4
6: err -66.95390862587533 Hz, inliers: 5
Estimated frequency offset: -66.95390862599015 Hz
/home/sunxc/code/python/miniconda3/envs/snappergps-envs/lib/python3.7/site-packages/autograd/tracer.py:48: RuntimeWarning: invalid value encountered in sqrt
  return f_raw(*args, **kwargs)
/home/sunxc/code/python/miniconda3/envs/snappergps-envs/lib/python3.7/site-packages/autograd/tracer.py:48: RuntimeWarning: divide by zero encountered in log
  return f_raw(*args, **kwargs)
Cannot calculate HDOP.
/home/sunxc/code/python/miniconda3/envs/snappergps-envs/lib/python3.7/site-packages/autograd/numpy/numpy_vjps.py:78: RuntimeWarning: invalid value encountered in true_divide
  defvjp(anp.log,    lambda ans, x : lambda g: g / x)
/home/sunxc/code/python/miniconda3/envs/snappergps-envs/lib/python3.7/site-packages/autograd/tracer.py:48: RuntimeWarning: invalid value encountered in remainder
  return f_raw(*args, **kwargs)
2120 20.031829280601926 110.40235092198098 0.0 inf
Uploaded record.
2121 19.90845894687515 110.31482566966899 0.0 inf
Uploaded record.
2122 20.006316334500248 110.27165954351719 0.0 inf
Uploaded record.
2123 19.97919230780266 110.41294086664357 0.0 inf
Uploaded record.
2124 20.089122813319456 110.33828422575117 0.0 inf
Uploaded record.
2125 20.072189799791367 110.30705149336089 0.0 inf
Uploaded record.
2126 20.089117878535024 110.3752298480076 0.0 inf
Uploaded record.
2127 19.995496099300684 110.2618169845695 0.0 inf
Uploaded record.
2128 20.053625540261443 110.34062045893297 0.0 inf
Uploaded record.
2129 20.059534296752773 110.33046307552041 0.0 inf
Uploaded record.
2130 20.003304182677304 110.33600622939063 0.0 inf
Uploaded record.
2131 19.95720348116739 110.43388782594245 0.0 inf
Uploaded record.
2132 20.055308299248328 110.2458745436694 0.0 inf
Uploaded record.
2133 20.08909868835221 110.43397043227974 0.0 inf
Uploaded record.
2134 19.916887806059414 110.39347077288208 0.0 inf
Uploaded record.
2135 20.027163734057225 110.27077495075834 0.0 inf
Uploaded record.
2136 20.05246184148958 110.34579410703113 0.0 inf
Uploaded record.
2137 20.088933871440332 110.26115311919153 0.0 inf
Uploaded record.
2138 19.98988766874073 110.3285177733937 0.0 inf
Uploaded record.
2139 19.979698388971297 110.35223629529784 0.0 inf
Uploaded record.
2140 20.08468506158049 110.30200383095105 0.0 inf
Uploaded record.
2141 20.03033570567731 110.3343793995447 0.0 inf
Uploaded record.
2142 19.915690517544807 110.28291771649842 0.0 inf
Uploaded record.
2143 20.012898573814397 110.34233138304035 0.0 inf
Uploaded record.
2144 19.908441645810704 110.25047768606609 0.0 inf
Uploaded record.
2145 20.005878580812663 110.3068937625168 0.0 inf
Uploaded record.
2146 19.99292921867784 110.40105217424713 0.0 inf
Uploaded record.
2147 20.010480690672196 110.30859933778925 0.0 inf
Uploaded record.
2148 20.08911712826579 110.37866262703372 0.0 inf
Uploaded record.
2149 19.99898693891227 110.37799856835592 0.0 inf
Uploaded record.
2150 19.908461256587877 110.31105644795684 0.0 inf
Uploaded record.
2151 19.90843486631602 110.24285058239583 0.0 inf
Uploaded record.
2152 19.90906260848121 110.40155727432098 0.0 inf
Uploaded record.
2153 19.90845742797702 110.38426298656101 0.0 inf
Uploaded record.
2154 20.08910373798354 110.25867208269354 0.0 inf
Uploaded record.
2155 19.90843486631602 110.43386141760418 0.0 inf
Uploaded record.
2156 20.04760346065098 110.39024496148186 0.0 inf
Uploaded record.
2157 19.925848193152255 110.42217598950884 0.0 inf
Uploaded record.
2158 19.982178627457273 110.24281291067064 0.0 inf
Uploaded record.
2159 19.908434866325337 110.43386141760418 0.0 inf
Uploaded record.
2160 19.998227405133438 110.33569003639472 0.0 inf
Uploaded record.
2161 19.92020662518349 110.30507341213145 0.0 inf
Uploaded record.
2162 20.08800694664834 110.38808538064374 0.0 inf
Uploaded record.
2163 19.998793000000003 110.338356 0.0 inf
Uploaded record.
2164 19.998793000000003 110.338356 0.0 inf
Uploaded record.
2165 19.992590169984062 110.3165662787929 0.0 inf
Uploaded record.
2166 19.90843486631602 110.43386141760418 0.0 inf
Uploaded record.
2167 19.90843486631602 110.43386141760418 0.0 inf
Uploaded record.
2168 19.90843486631602 110.43386141760418 0.0 inf
Uploaded record.
2169 20.058389604503482 110.43395188432437 0.0 inf
Uploaded record.
2170 19.90843486631602 110.43386141760418 0.0 inf
Uploaded record.
2171 19.908437716190868 110.4338582257289 0.0 inf
Uploaded record.
2172 19.92064771325026 110.43386578529686 0.0 inf
Uploaded record.
2173 19.908460344692404 110.34586055033512 0.0 inf
Uploaded record.
2174 19.979975402931032 110.3049742986424 0.0 inf
Uploaded record.
Updating upload status to complete.
Then I checked on the web page, and it said I had located 56 snapshots, but it didn't show anything on the map：
![image](https://github.com/user-attachments/assets/dd45023b-0e86-44bf-bcba-de52f2215d48)

**Comment by sunmouren** [comment]
Only 55 snapshots were processed because of my time zone problem, which resulted in only 55 use_idx. Now all snapshots have been processed, but the web app still does not display. 
![image](https://github.com/user-attachments/assets/71c8d2c8-db5d-4f63-bba9-40ea1faab7d0)
I have replaced the mapboxAccessToken. What scopes does the mapboxAccessToken require?
![image](https://github.com/user-attachments/assets/09af8ef6-bfbf-47b5-91f8-1f10274273d5)

**Comment by JonasBchrt** [comment]
I have ticked all the boxes for public scopes, but doubt it's necessary.

**Comment by JonasBchrt** [comment]
But the reason why there is nothing displayed is that the code is not confident in 100% (all) of your fixes. Fixes without confidences are hidden by default. You can show them by clicking "Markers" in the drop-down menu on the top-right, but they are likely garbage.

**Comment by JonasBchrt** [comment]
The back-end fails to calculate any reliable fixes, which is surprising, given that the same dataset seemed to have been successful for you when using the online service. The frequency offset estimation also seemed to have worked.

**Comment by JonasBchrt** [comment]
Some things to check:
- Start location within 10 km of actual location.
- Start time correct.
- Any code or package changes?

**Comment by sunmouren** [comment]
> Some things to check:
> 
> * Start location within 10 km of actual location.
> * Start time correct.
> * Any code or package changes?

I use centos7.9 operating system, and use conda to create a python3.7.16 virtual environment. When I execute pip install -r requirements.txt to install third-party libraries, rockhound fails, so I use conda install rockhound separately. This is my pip freese:

> affine==2.4.0
aiohttp==3.8.6
aiosignal==1.3.1
APScheduler==3.6.3
async-timeout==4.0.3
asynctest==0.13.0
attrs==24.2.0
autograd==1.6.2
autoptim==0.3
backports.zoneinfo==0.2.1
certifi @ file:///croot/certifi_1671487769961/work/certifi
cffi==1.15.1
cftime @ file:///tmp/build/80754af9/cftime_1638345281172/work
charset-normalizer==3.4.0
click==8.1.7
click-plugins==1.1.1
cligj==0.7.2
cloudpickle==2.2.1
cryptography==44.0.0
dask==2022.2.0
frozenlist==1.3.3
fsspec==2023.1.0
future==1.0.0
http-ece==1.2.1
idna==3.10
importlib-metadata==6.7.0
joblib==1.3.2
locket==1.0.0
mkl-fft==1.3.1
mkl-random @ file:///tmp/build/80754af9/mkl_random_1626179032232/work
mkl-service==2.4.0
multidict==6.0.5
netCDF4 @ file:///croot/netcdf4_1673455456943/work
numpy @ file:///opt/conda/conda-bld/numpy_and_numpy_base_1653915516269/work
packaging==24.0
pandas==1.3.5
partd==1.4.1
platformdirs==4.0.0
pooch==1.8.2
psycopg2-binary==2.9.9
py-vapid==1.9.2
pycparser==2.21
PyGeodesy==24.11.11
pymap3d==2.9.1
pyparsing==3.1.4
python-dateutil==2.9.0.post0
python-telegram-bot==13.5
pytz==2024.2
pywebpush==2.0.3
PyYAML==6.0.1
rasterio==1.2.10
requests==2.31.0
rockhound==0.2.0
scikit-learn==0.24.0
scipy==1.7.3
six @ file:///tmp/build/80754af9/six_1644875935023/work
snuggs==1.4.7
threadpoolctl==3.1.0
toolz==0.12.1
tornado==6.2
typing_extensions==4.7.1
tzlocal==5.1
urllib3==2.0.7
xarray==0.17.0
yarl==1.9.4
zipp==3.15.0

Here is the requirements.txt for your github repository:

> numpy
python-telegram-bot == 13.5
pymap3d
autograd
psycopg2-binary
pywebpush
scipy
scikit-learn == 0.24.0
autoptim
PyGeodesy
rockhound
xarray == 0.17.0


Can you provide the pip freeze you actually deployed and installed?

**Comment by JonasBchrt** [comment]
This would be an exemplary output of `pip freeze`:
```
affine==2.4.0
aiohttp==3.8.6
aiosignal==1.3.1
APScheduler==3.6.3
async-timeout==4.0.3
asynctest==0.13.0
attrs==23.2.0
autograd==1.6.2
autoptim==0.3
backports.zoneinfo==0.2.1
certifi==2024.6.2
cffi==1.15.1
cftime==1.6.2
charset-normalizer==3.3.2
click==8.1.7
click-plugins==1.1.1
cligj==0.7.2
cloudpickle==2.2.1
cryptography==42.0.8
cycler==0.11.0
dask==2022.2.0
fonttools==4.38.0
frozenlist==1.3.3
fsspec==2023.1.0
future==1.0.0
http-ece==1.2.0
idna==3.7
importlib-metadata==6.7.0
joblib==1.3.2
kiwisolver==1.4.5
locket==1.0.0
matplotlib==3.5.3
multidict==6.0.5
netCDF4==1.6.5
numpy==1.21.6
packaging==24.0
pandas==1.3.5
partd==1.4.1
Pillow==9.5.0
platformdirs==4.0.0
pooch==1.8.2
psycopg2-binary==2.9.9
py-vapid==1.9.1
pycparser==2.21
PyGeodesy==24.6.9
pymap3d==2.9.1
pyparsing==3.1.2
python-dateutil==2.9.0.post0
python-telegram-bot==13.5
pytz==2024.1
pywebpush==2.0.0
PyYAML==6.0.1
rasterio==1.2.10
requests==2.31.0
rockhound==0.2.0
scikit-learn==0.24.0
scipy==1.7.3
six==1.16.0
snuggs==1.4.7
threadpoolctl==3.1.0
toolz==0.12.1
tornado==6.2
typing_extensions==4.7.1
tzlocal==5.1
urllib3==2.0.7
xarray==0.17.0
yarl==1.9.4
zipp==3.15.0
```
Running Ubuntu 22.04.5 LTS.

**Comment by JonasBchrt** [comment]
Here's the navigation data used on the server, for comparison (SnapperGPS format):
[2024_340_.zip](https://github.com/user-attachments/files/18135140/2024_340_.zip)

**Comment by sunmouren** [comment]
> Here's the navigation data used on the server, for comparison (SnapperGPS format):
> [2024_340_.zip](https://github.com/user-attachments/files/18135140/2024_340_.zip)
> 

Thanks, I will try again next Monday as I don't have a computer right now.

**Comment by sunmouren** [comment]
> Some things to check:
> - Start location within 10 km of actual location.
> - Start time correct.
> - Any code or package changes?

I checked all of these, and they are all ok. I checked the version of the repository with pip freeze, and it is the same as yours. I also checked the code with git version changes, and it shows no changes. Finally, it replaced the navigation data you gave, but its calculation result is still not good as before. It is so confusing. Did you clone the code directly from the git repository? Or did you make some changes?

**Comment by JonasBchrt** [comment]
Hi @sunmouren, yeah, the code on the servers is from Github. No changes in the `core` directory. I only adjust the code in `web_app/processing` to the server environment. (Settings in `config.py` and some automated crash reporting in the other files.)

**Comment by JonasBchrt** [comment]
I don't have an idea right know where your problem may stem from, but will think about it and come back to you with any ideas.

**Comment by JonasBchrt** [comment]
Did you ever manage to successfully run https://github.com/JonasBchrt/snapshot-gnss-algorithms?

**Comment by JonasBchrt** [comment]
You could also try to run one of the test datasets with your local copy of the code: https://ora.ox.ac.uk/objects/uuid:cf48364c-c837-4dbd-b615-982eb6eb7aff

**Comment by JonasBchrt** [comment]
Just use Oxford latitude/longitude as start location.

**Comment by sunmouren** [comment]
> Did you ever manage to successfully run https://github.com/JonasBchrt/snapshot-gnss-algorithms?

I have not run it before. I have tested snappergps-backend on ubuntu22.04 and centos 7.9 respectively. The results are the same. The following error message is prompted during the operation:

![image](https://github.com/user-attachments/assets/5ba15f50-804c-485c-935d-18f93c3916b1)

![image](https://github.com/user-attachments/assets/50ff1394-cc9f-47b1-be44-c3aaa2f73995)

I thought about it, could you recalculate the snapshot I uploaded to your online server before, and see if the calculation process outputs similar logs as above? 

[24776C025FE526AC_20241205_075530_000.json](https://github.com/user-attachments/files/18163693/24776C025FE526AC_20241205_075530_000.json)

**Comment by sunmouren** [comment]
> You could also try to run one of the test datasets with your local copy of the code: https://ora.ox.ac.uk/objects/uuid:cf48364c-c837-4dbd-b615-982eb6eb7aff

Here are the results. HDOP cannot be calculated, and Estimates without confidence: 100.0%. The only difference I can think of from your online server is the time zone.

![image](https://github.com/user-attachments/assets/7fe01f2d-7b1a-461e-aaa2-94ab6f2edaa1)

![image](https://github.com/user-attachments/assets/d3250c12-8104-463e-80fa-3e43b98a6a77)

**Comment by JonasBchrt** [comment]
I will have a look later this week. I am quite keen on making sure that people can run their own SnapperGPS server if they wish to.

Just regarding the warning messages:
I think the invalid value warning shows sometimes up when the algorithm is not able to successfully calculate a reliable fix. Similarly, the "Cannot calculate HDOP" message shows up when the algorithm is not able to calculate a finite HDOP for a fix, i.e., the fix is likely bad. In general, you may see both messages even during "normal" server operation, e.g., if a few fixes were recorded indoors. But for you, it pops up for every fix, apparently.
So, in summary, these messages are symptoms of a problem, but do not really give a clue regarding the problem's root cause.

**Comment by sunmouren** [comment]
> I will have a look later this week. I am quite keen on making sure that people can run their own SnapperGPS server if they wish to.
> 
> Just regarding the warning messages:
> I think the invalid value warning shows sometimes up when the algorithm is not able to successfully calculate a reliable fix. Similarly, the "Cannot calculate HDOP" message shows up when the algorithm is not able to calculate a finite HDOP for a fix, i.e., the fix is likely bad. In general, you may see both messages even during "normal" server operation, e.g., if a few fixes were recorded indoors. But for you, it pops up for every fix, apparently.
> So, in summary, these messages are symptoms of a problem, but do not really give a clue regarding the problem's root cause.

Is it possible that the database version may affect the calculation results? The database version I am currently using is not 14.7. Or will the computer's local time zone not be UTC affect the calculation results? During the calculation process, will the code use the local time zone?

**Comment by JonasBchrt** [comment]
A bit unrelated, but I just set up the back-end from scratch using the https://github.com/SnapperGPS/snappergps-backend repository and precisely following the readme.
Processing data works fine; so, I don't think there is a fundamental issue with the repository, but instead some interplay with your specific setup.

I used Ubuntu 23.xx and postgres 14.13 for this specific test, but I doubt any of this is a game changer.

I did run the backend in quite a few different timezones in the past; so, I don't expect any issues stemming from that - albeit, I didn't check that today.

In contrast, the database itself has not moved timezones as much... Maybe there's something odd. But the dates are all stored as `timestamp(3) without time zone`, which should avoid problems.

Would it be hard for you to repeat a test with all your hosts (database, backend, webserver) set to UTC?

**Comment by sunmouren** [comment]
> A bit unrelated, but I just set up the back-end from scratch using the https://github.com/SnapperGPS/snappergps-backend repository and precisely following the readme. Processing data works fine; so, I don't think there is a fundamental issue with the repository, but instead some interplay with your specific setup.
> 
> I used Ubuntu 23.xx and postgres 14.13 for this specific test, but I doubt any of this is a game changer.
> 
> I did run the backend in quite a few different timezones in the past; so, I don't expect any issues stemming from that - albeit, I didn't check that today.
> 
> In contrast, the database itself has not moved timezones as much... Maybe there's something odd. But the dates are all stored as `timestamp(3) without time zone`, which should avoid problems.
> 
> Would it be hard for you to repeat a test with all your hosts (database, backend, webserver) set to UTC?

Processing data is working fine：
![image](https://github.com/user-attachments/assets/f67d902f-6713-42b8-b653-2b892a9a2acc)
The problem is the time zone. The time zone of the entire environment (including postgresql, snappergps-app, snappergps-backend) must be UTC+00:00. I checked the time zones of all my previous environments. The time zones of postgresql and snappergps-app are UTC+08:00 (the computer is on the local computer), and the time zone of snappergps-backend is UTC+00:00 (the computer is on the cloud). I set the time zone of the computer where postgresql and snappergps-app are located to UTC+00:00, then cleared all database data, and re-uploaded the data, and it worked normally. But the dates are all stored as timestamp(3) without time zone in postgresql database, so I think the problem may be in the time zone of snappergps-app and it is most likely a time zone problem of nodejs.

**Comment by JonasBchrt** [comment]
Great work, thanks for checking.
Yes, I think it's the node.js runtime that needs to be on a server with UTC timezone.
I have not tested it yet, but I am pretty sure the mistake lies in [dbFunctions.js](https://github.com/SnapperGPS/snappergps-app/blob/main/js/dbFunctions.js) in combination with the postgres schema. The calls to `getHours()` [here](https://github.com/SnapperGPS/snappergps-app/blob/13c745600f56a3971c14b010878fb2c4b7c02f5b/js/dbFunctions.js#L120) return [local time](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getHours). We probably want [`getUTCHours()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getUTCHours). Same for all the other calls of functions of the `Date` class in `dbFunctions.js`. Alternatively, use timezone aware timestamps in the postgres schema and ensure that the postgres DB runs on a server with the same timezone as the node.js runtime. But this is probably the worse fix of the two.

**Comment by JonasBchrt** [comment]
I don't think there is an issue with the Python back-end.
So, I made another issue here: https://github.com/SnapperGPS/snappergps-app/issues/9
Mainly for myself to keep track of it.

**Comment by JonasBchrt** [comment]
Addressed in commit https://github.com/SnapperGPS/snappergps-app/commit/5f2f42da5c3ac4a4bcead3dd250be1b63a22e188.
